Lesson 5-7: Driver for character device

D-Alexandr opened the pull request:
Added Makefile and driver for character device.

@oleg-h2o , @galkinletter please check my homework

D-Alexandr committed:
Lesson 5-7 Part 1: Added driver interface and makefile

A structure with pointers to the functions of working with the driver has been created. Written makefile with two target: clean, all

Signed-off-by: Alexandr Datsenko <datsenkoalexander@gmail.com>

D-Alexandr committed:
Lesson 5-7 Part 1: Added loading and unloading module

Added function of loading and unloading module. Added parameter buffer size. The minimum size of the buffer is 1000 bits.

Signed-off-by: Alexandr Datsenko <datsenkoalexander@gmail.com>

D-Alexandr committed:
Lesson 5-7 Part 1: Added driver functionality read and write

Added function to write and read a stream of characters from the driver buffer. Added check for minimum driver buffer size.

Signed-off-by: Alexandr Datsenko <datsenkoalexander@gmail.com>

galkinletter commented on 02-KernelModules_CharacterDevice/chrdev_driver.c:
```diff
@@ -0,0 +1,188 @@
+	}
+	cdev_init(cdev, &chrdev_fops);
+	cdev->owner   = THIS_MODULE;
+	if (cdev_add(cdev, dev, chrdev_nr_devs)) { 
```
```suggestion
	if (cdev_add(cdev, dev, chrdev_nr_devs)) { 
			printk(KERN_WARNING "chrdev: cdev_add error \n");
		}
		printk(KERN_INFO "chrdev: %d:%d \n", chrdev_major, chrdev_minor);
	
```

printk(KERN_WARNING  "chrdev: cdev_add error \n");

Please do it error handling  for this.

Think of fatal errors and their handling.

galkinletter reviewed (changes requested):
Please do it error handling for some events

Think of fatal errors and their handling.

D-Alexandr committed:
Lesson 5-7 Part 1: Added error messages and info to handlers

Added error handler to cdev_add function. Written more detailed error messages. Fixed code style

Signed-off-by: Alexandr Datsenko <datsenkoalexander@gmail.com>

D-Alexandr commented:
Fixed. @galkinletter  please check

galkinletter commented on 02-KernelModules_CharacterDevice/chrdev_driver.c:
```diff
@@ -0,0 +1,197 @@
+		chrdev_buffer = kzalloc(chrdev_buffer_size * sizeof (*chrdev_buffer), GFP_KERNEL);
+	if (!chrdev_buffer) {
+		result = -ENOMEM;
+		printk(KERN_ERR "chrdev: chrdev_buffer is out of memory");
```
```suggestion
		printk(KERN_ERR "chrdev: chrdev_buffer is out of memory");
```
Think about freeing up the required resources after errors

galkinletter commented on 02-KernelModules_CharacterDevice/chrdev_driver.c:
```diff
@@ -0,0 +1,197 @@
+	 if (!cdev) {
+		result = -ENOMEM;
+		printk(KERN_ERR "chrdev: cdev is out of memory");
+		goto fail; 
```
```suggestion
		goto fail; 
```
Think about freeing up the required resources after errors

galkinletter reviewed (changes requested):
Think about freeing up the required resources after errors

galkinletter commented:
Think about freeing up the required resources after errors

D-Alexandr committed:
Lesson 5-7 Part 2: Added procfs interface to control driver

Added proc_chrdev_read function which shows the used volume and buffer size. Fixed error handler

Signed-off-by: Alexandr Datsenko <datsenkoalexander@gmail.com>

D-Alexandr committed:
Lesson 5-7 Part 2: Added sysfs interface to control driver

Added function chrdev_cdev_store, which clears the driver buffer. Added removal of driver folders after execution rmmod

Signed-off-by: Alexandr Datsenko <datsenkoalexander@gmail.com>

D-Alexandr committed:
Lesson 5-7 Part 3: Added test script and instruction

Added a test script with different scenarios test the driver. Written instructions for working with the test script.
Modified module completion function

Signed-off-by: Alexandr Datsenko <datsenkoalexander@gmail.com>

oleg-h2o commented on 02-KernelModules_CharacterDevice/chrdev_driver.c:
```diff
@@ -0,0 +1,298 @@
+static ssize_t chrdev_cdev_store(struct kobject *kobj, struct kobj_attribute *attr, const char *buf, size_t count)
+{
+	int clean_flag = 0;
+    sscanf(buf, "%d", &clean_flag);
```
Here're issues with indentation.

oleg-h2o reviewed (changes requested):
Pl fix issues with indentation (checkpatch or whatever...)

D-Alexandr committed:
Lesson 5-7 Part 3: Fixed issues with indentation

Corrected code style using checkpatch. Added check for sscanf

Signed-off-by: Alexandr Datsenko <datsenkoalexander@gmail.com>

galkinletter commented:
fixed is correct
